logicalplan/regex_matchers.go

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/thanos-community/promql-engine/parser"
+	"regexp"
+)
+
+var hasRegexString = regexp.MustCompile(`^[^\\\\[{()^$.*?|+]*$`)
```

Comment:
Is the regular expression here necessary? Probably it would be safer to use `regexp/syntax` here, parse the regular expression and check that the topmost opcode is "OpLiteral"?

Reply:
Or is parsing the expensive part?

Reply:
compiling a regular expression is generally faster than parsing the regular expression because `regexp.MustCompile` function caches the compiled regular expression, allowing subsequent uses of the regular expression to be faster.

Reply:
I meant to check if the incoming regular expression is a literal; but probably parsing here is just making the problem worse

Reply:
couldnt we do this without the regex and a simple `strings.ContainsAny(regexFromLabel, specialRegexChars)` ?

Reply:
No, we don't want change the query if input string contains valid expression. If you have better idea please feel free to share.

Reply:
Wouldn't that be equivalent though? I meant if it doesn't contain special regex chars it's probably a literal and can be optimized with this, right?

Reply:
ah yes